src/HashStore.hs

```diff
@@ -1,16 +1,49 @@
 module HashStore
-       ( hashFile
+       ( hashStore
        ) where
 
 import Crypto.Hash.BLAKE2.BLAKE2b (hash)
-import Data.ByteString.Char8 (ByteString, unpack)
+import Data.ByteString.Char8 (ByteString)
+import System.Directory (createDirectoryIfMissing, doesFileExist)
+import System.FilePath ((</>))
 
+import qualified Data.ByteString.Char8 as BS
+
+-- QUESTION: does it make sense to create `newtype Hash = Hash ByteString`?
 -- | Perform 'ByteString' hashing.
 hashContent :: ByteString -> ByteString
 hashContent = hash 64 mempty
 {-# INLINE hashContent #-}
 
 -- | Build the new 'FilePath' adding the hash of the provided content.
 hashFile :: ByteString -> FilePath -> FilePath
-hashFile content file = unpack (hashContent content) ++ "-" ++ file
+hashFile content file = BS.unpack (hashContent content) ++ "-" ++ file
 {-# INLINE hashFile #-}
+
+{- | @'hashStore' storePath action (name, content)@ computes @hash@ of @content@
+and performs @action@ if file @storePath '</>' (hash ++ "-" ++ name)@ doesn't
```

Comment:
Probably it's better to write it in some pseudo path instead of real code. Something like `@storePath/hash-name@` will be more readable, in my opinion.